php/libraries/Utility.class.inc

```diff
@@ -933,5 +935,100 @@ class Utility
         }
         return $languages;
     }
+
+    /**
+     * Updates the password for this user into the appropriate field in the
+     * database.
+     *
+     * @param \User     $user     The plain text password to be hashed and saved.
+     * @param \Password $password The plain text password to be hashed and saved.
+     * @param ?DateTime $expiry   When the password will expire. Can be set to a
+     *                            date in the past in order to create a one-time
+     *                            password.
+     *
+     * @return void
+     */
+    static function updatePassword(
+        \User $user,
+        \Password $password,
+        ?\DateTime $expiry = null
+    ): void {
+        // Set default expiry date value to 6 months in the future.
+        $expiry = $expiry ?? new \DateTime(
+            date('Y-m-d', strtotime('+6 months'))
+        );
+        error_log($expiry->format(self::MYSQL_DATE_FORMAT));
+
+        // Update the users table to change $user's pasword to $password.
+        \NDB_Factory::singleton()->database()->update(
+            'users',
+            array(
+             'Password_hash'   => $password,
+             'Password_expiry' => $expiry->format(self::MYSQL_DATE_FORMAT),
+            ),
+            array(
+             'UserID' => $user->getUsername(),
+            )
+        );
+    }
+
+    /**
+     * Determines if the supplied password parameter is different than a User's
+     * current password (the one stored in the database).
+     *
+     * @param \User  $user              User whose password may have changed
+     * @param string $plaintextPassword The new desired password string entered
+     *                                  by the user.
+     *
+     * @return bool  true if the password has changed, false otherwise.
+     */
+    static function passwordChanged(
```

Comment:
The `!=` operator would replace that function.
something like `$user->getPassword() != (new \Password($value))`

Reply:
In the PR description I linked another PR's of Dave's where he suggested removing all the password functionality from the User class in order to more accurately represent a user.

By the `!=` operator are you talking about the old `save()` function and the `User` class? I was always confused by what that did and where the value came from.

Also, even with those functions there was a `passwordChanged` function needed in both the Edit User and My Preferences code. 

I think it makes more sense to do away with it to simplify the User class as well as to reduce code reusage by replacing `save()` and the `passwordChanged` functions with this one.

Reply:
@johnsaigle It is not related to the save function.

the `!=` operator could be used to compare two `Password` objects. 
```
$userpassword = $user->getPassword();
$newpassword = new \Password($value);
if ($userpassword != $newpassword) {
    echo "Those are different password.";
} else {
    echo "Those are the same password.";
}
```

Reply:
The `__toString()` function of the `Password` class is the hash of the password, not the plaintext password. I chose to do this so that there's no risk of the plaintext getting exposed anywhere.

Also there isn't a `getPassword()` function.

Reply:
@johnsaigle keeping the hash is great and do not prevent from comparing 2 password using `==`. 
But what will prevent it is the random salt used by password_hash.